File: Chapter 02/JS/section_2_7_1_weighted_decision_model/decisionMaker.js
/**
 * Implements a Weighted Decision Model to choose the best option from a set of choices.
 * This model provides a quantitative and data-driven way to make architectural decisions.
 * 
 * @param {Array<Option>} options - A list of options to evaluate. Each option has scores for various criteria.
 * @param {Object.<string, number>} weights - An object where the key is the criterion name and the value is its importance (weight).
 * @returns {{bestOption: string, rationale: string}} An object containing the name of the best option and a string explaining the rationale.
 */
function pickOption(options, weights) {
    let bestOption = null;
    let highestScore = -Infinity;
    const details = [];

    for (const opt of options) {
        // THE CORE LOGIC: Calculate the weighted score for this option.
        // For each criterion (e.g., "performance"), we multiply the option's
        // score for that criterion (e.g., 4/5) by the weight we've assigned to that
        // criterion (e.g., 60% importance) and sum the results.
        // Formula: FinalScore = (Score_A * Weight_A) + (Score_B * Weight_B) + ...
        let score = 0;
        for (const [key, weight] of Object.entries(weights)) {
            score += (opt.scores[key] || 0) * weight;
        }
        details.push(`${opt.name}: ${score.toFixed(2)}`);

        if (score > highestScore) {
            highestScore = score;
            bestOption = opt;
        }
    }

    // The rationale provides a transparent explanation for the decision.
    const weightsString = JSON.stringify(weights).replace(/"/g, "'");
    const rationale = `Scores: ${details.join(' | ')}\\n -> Based on weights ${weightsString}, we pick **${bestOption.name}**.`;
    
    return { bestOption: bestOption.name, rationale };
}

module.exports = { pickOption };

File: Chapter 02/Python/section_2_7_1_weighted_decision_model/decision_maker.py
from typing import Dict, Tuple, List
from option import Option

def pick_option(options: List[Option], weights: Dict[str, float]) -> Tuple[str, str]:
    """
    Implements a Weighted Decision Model to choose the best option from a set of choices.
    This model provides a quantitative and data-driven way to make architectural decisions.

    Args:
        options: A list of options to evaluate. Each option has scores for various criteria.
        weights: A dictionary where the key is the criterion name and the value is its importance (weight).

    Returns:
        A tuple containing the name of the best option and a string explaining the rationale.
    """
    best_option = None
    highest_score = float("-inf")
    details = []

    for opt in options:
        # THE CORE LOGIC: Calculate the weighted score.
        # Formula: FinalScore = (Score_A * Weight_A) + (Score_B * Weight_B) + ...
        score = sum(opt.scores.get(k, 0) * weights.get(k, 0.0) 
                    for k in weights)
        details.append(f"{opt.name}: {score:.2f}")

        if score > highest_score:
            highest_score = score
            best_option = opt

    # The rationale provides a transparent explanation for the decision.
    rationale = f"Scores: {' | '.join(details)}\\n -> Based on weights {weights}, we pick **{best_option.name}**."
    return best_option.name, rationale